common/verify/proof.py

```
@@ -0,0 +1,226 @@
+    is_leaf = (flag & 2) != 0
+    is_odd = (flag & 1) != 0
+    nibbles: list[int] = []
+    if is_odd:
+        nibbles.append(encoded[0] & 0x0F)
+    for byte in encoded[1:]:
+        nibbles.append(byte >> 4)
+        nibbles.append(byte & 0x0F)
+    return nibbles, is_leaf
+
+
+def _follow_reference(ref: Any, proof_by_hash: dict[bytes, bytes]) -> Any:
```

Ruff (0.15.12)

[warning] Dynamically typed expressions (typing.Any) are disallowed in `ref`

(ANN401)